[Media] Destroy media player on fatal playback error

eocanha opened the pull request:
#### 1c9cc6e485dc2ebd5604ba671a580ba5e7a9c7b7
<pre>
[Media] Destroy media player on fatal playback error
<a href="https://bugs.webkit.org/show_bug.cgi?id=242395">https://bugs.webkit.org/show_bug.cgi?id=242395</a>

Reviewed by Xabier Rodriguez-Calvar.

Under some circumstances, the SourceBuffer.append() can generate an EOS internally in glib ports, which
reaches HTMLMediaElement as DecodeError. This triggers detachMediaSource() while the player is still playing
fine (it&apos;s not aware of any issue). Eventually, the player reaches EOS when running out of samples and the
playback is ended. The JavaScript app may try to resume video with video.play(), triggering seek(0), which
triggers other internal operations. One of those operations was a MediaPlayerPrivateGStreamerMSE::sourceSetup()
call that was trying to setPrivateAndOpen() on the MediaSource that was already detached and this was crashing.
The crash itself was fixed in <a href="https://bugs.webkit.org/show_bug.cgi?id=220091">https://bugs.webkit.org/show_bug.cgi?id=220091</a> by reopening MediaSource on
load() only, but the player private still keeps this detached MediaSource that may cause different issues.

The proper way to avoid this invalid state where a reference to MediaSource is kept by the player private is
to ensure that the player is destroyed on error. It will be recreated again if needed if a fresh src is set in
the future.

Original author: Andrzej Surdej &lt;Andrzej_Surdej@comcast.com&gt;

See: <a href="https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/899">https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/899</a>

* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::mediaLoadingFailedFatally): Clear media player on error.

Canonical link: <a href="https://commits.webkit.org/253522@main">https://commits.webkit.org/253522@main</a>
</pre>

calvaris reviewed (approved):
Let's wait for some time to get more comments just in case.

eocanha committed:
[Media] Destroy media player on fatal playback error
https://bugs.webkit.org/show_bug.cgi?id=242395

Reviewed by Xabier Rodriguez-Calvar.

Under some circumstances, the SourceBuffer.append() can generate an EOS internally in glib ports, which
reaches HTMLMediaElement as DecodeError. This triggers detachMediaSource() while the player is still playing
fine (it's not aware of any issue). Eventually, the player reaches EOS when running out of samples and the
playback is ended. The JavaScript app may try to resume video with video.play(), triggering seek(0), which
triggers other internal operations. One of those operations was a MediaPlayerPrivateGStreamerMSE::sourceSetup()
call that was trying to setPrivateAndOpen() on the MediaSource that was already detached and this was crashing.
The crash itself was fixed in https://bugs.webkit.org/show_bug.cgi?id=220091 by reopening MediaSource on
load() only, but the player private still keeps this detached MediaSource that may cause different issues.

The proper way to avoid this invalid state where a reference to MediaSource is kept by the player private is
to ensure that the player is destroyed on error. It will be recreated again if needed if a fresh src is set in
the future.

Original author: Andrzej Surdej <Andrzej_Surdej@comcast.com>

See: https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/899

* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::mediaLoadingFailedFatally): Clear media player on error.

Canonical link: https://commits.webkit.org/253522@main
